[DOC-9828] 23.1.17 release notes (#18387)
cockroachdb/docs: src/current/_includes/releases/v23.1/v23.1.17.md

## v23.1.17

Release Date: March 19, 2024

{% include releases/release-downloads-docker-image.md release=include.release %}

<h3 id="v23-1-17-security-updates">Security updates</h3>

> - A user with the `VIEWACTIVITY` privilege can now request a statement bundle from the [**Statements** page](https://www.cockroachlabs.com/docs/v23.1/ui-statements-page) in the DB Console or with the internal builtin function `crdb_internal.requests_statement_bundle`. [#119638][#119638]

Reviewer: 'the VIEWACTIVITY privilege' could be a link to https://www.cockroachlabs.com/docs/v23.1/security-reference/authorization#supported-privileges

> - Clusters using [Cluster Single Sign-on (SSO) with JSON web tokens (JWTs)](https://www.cockroachlabs.com/docs/v23.1/sso-sql) can now optionally fetch signing keys from configured issuers instead of configuring static signing keys for each issuer. When the new cluster setting `server.jwt_authentication.jwks_auto_fetch.enabled` is set to `true`, signing keys are automatically fetched from the issuer using metadata published in its OpenID configuration. In this case, static signing keys in `server.jwt_authentication.jwks` are ignored. When automatic fetching is enabled, there may be a slight increase in network latency for each JWT authentication request, proportional to the latency between the cluster and the issuer's endpoint. [#120063][#120063]

Reviewer: "cluster setting" could be a link to https://www.cockroachlabs.com/docs/v23.1/cluster-settings

> - DB Console cookies are marked `Secure` for the browser when the cluster is running in secure mode. [#119262][#119262]

Reviewer: "DB Console" could be a link to https://www.cockroachlabs.com/docs/v23.1/ui-overview

> - The DB Console `session` cookie is now marked `HttpOnly` to prevent it from being read by any Javascript code. [#119262][#119262]

Reviewer: "DB Console" could be a link to https://www.cockroachlabs.com/docs/v23.1/ui-overview

<h3 id="v23-1-17-sql-language-changes">SQL language changes</h3>

> - The new [cluster setting](https://cockroachlabs.com/docs/v23.1/cluster-settings) `server.max_open_transactions_per_gateway`, when set to a non-negative value, prevents users without the `admin` role from executing a query if more than this number of transactions are open on the current gateway node. [#118949][#118949]

Reviewer: 'admin role' could be a link to https://www.cockroachlabs.com/docs/v23.1/security-reference/authorization#admin-role

> - The new [cluster setting](https://cockroachlabs.com/docs/v23.1/cluster-settings) `server.max_open_transactions_per_gateway`, when set to a non-negative value, prevents users without the `admin` role from executing a query if more than this number of transactions are open on the current gateway node. [#118949][#118949]

Reviewer: 'gateway node' could be a link to https://www.cockroachlabs.com/docs/v23.1/architecture/life-of-a-distributed-transaction#gateway

> - The new [cluster setting](https://cockroachlabs.com/docs/v23.1/cluster-settings) `server.max_open_transactions_per_gateway`, when set to a non-negative value, prevents users without the `admin` role from executing a query if more than this number of transactions are open on the current gateway node. [#118949][#118949]

Reviewer: 'transactions' could be a link to https://www.cockroachlabs.com/docs/v23.1/transactions

> - Added support for [index hinting](https://www.cockroachlabs.com/docs/v23.1/table-expressions#force-index-selection) with `INSERT` and `UPSERT` statements. This allows `INSERT ... ON CONFLICT` and `UPSERT` queries to use index hints in the same way as `UPDATE` and `DELETE` statements. [#119600][#119600]

Reviewer: mentions of INSERT, UPSERT, UPDATE, DELETE could all be links to those statements' doc pages

<h3 id="v23-1-17-operational-changes">Operational changes</h3>

- You can now pass the `--include-range-info` flag when [generating a diagnostic bundle](https://www.cockroachlabs.com/docs/v23.1/cockroach-debug-zip) to include problem ranges. [#119233][#119233]
- In unredacted diagnostic bundles, two columns have been added to the `crdb_internal.transaction_contention_events` table:
    - `waiting_stmt_query`: the query of the statement that is waiting to run.
    - `blocking_txn_queries_unordered`: an unordered list of the blocking transaction's queries. [#119422]

>     [#119422]

Reviewer: looks like this markdown link tag got bumped to a new line,  unintentional?  usual format is `[#xxx][#xxx]` on same line

Reviewer: oh yeah in the rendered preview it's a second duplicate link on a separate line

<h3 id="v23-1-17-db-console-changes">DB Console changes</h3>

- The [**Statements** page](https://www.cockroachlabs.com/docs/v23.1/ui-statements-page) now always shows the entire selected period, instead of only the period that contains data. [#118811][#118811]
- The [**Overload Dashboard**](https://www.cockroachlabs.com/docs/v23.1/ui-overload-dashboard) page now includes two additional graphs:
    - **Elastic CPU Utilization**: Shows the actual CPU used for elastic work compared with the configured limit.
    - **Elastic CPU Exhausted Duration Per Second**: Shows how much time, in milliseconds, that elastic work has been subject to CPU exhaustion. [#118908][#118908]

<h3 id="v23-1-17-bug-fixes">Bug fixes</h3>

> - Fixed a bug where creating a [changefeed](https://www.cockroachlabs.com/docs/v23.1/change-data-capture-overview) that targeted tables with a zero-scale `DECIMAL(n)` column, `format='avro'`, and `diff` would cause a panic. [#118848][#118848]

Reviewer: DECIMAL could be a link to the docs at https://www.cockroachlabs.com/docs/v23.1/decimal

    1. A transaction commits and removes its [transaction record](https://www.cockroachlabs.com/docs/v23.1/architecture/transaction-layer#transaction-records) before its [intent](https://www.cockroachlabs.com/docs/v23.1/architecture/transaction-layer#writing) resolution is applied on the follower.
    1. The follower's [closed timestamp](https://www.cockroachlabs.com/docs/v23.1/architecture/transaction-layer#closed-timestamps) has advanced past the transaction commit timestamp.
    1. The rangefeed attempts to push the transaction to a new timestamp (at least 10 seconds after the transaction began).
    1. This may cause the rangefeed to prematurely emit a checkpoint before emitting writes at lower timestamps, which in turn may cause the [changefeed](https://www.cockroachlabs.com/docs/v23.1/change-data-capture-overview#how-does-an-enterprise-changefeed-work) to drop these events entirely, never emitting them.

>     [#118415][#118415]

Reviewer: intended to be on separate line? seems to work ok in the rendered preview

> - Fixed a bug introduced in v22.2 where queries issued by the TTL jobs did not use optimal plans. [#118496][#118496]

Reviewer: 'TTL jobs' could be a link to https://www.cockroachlabs.com/docs/v23.1/row-level-ttl#view-scheduled-ttl-jobs

> - Fixed a bug introduced in v23.1.0 where a decommissioning replica that is part of a mis-replicated range could get stuck on a rebalance operation that was falsely determined to be unsafe. [#118437][#118437]

Reviewer: 'decommissioning' could be a link to https://www.cockroachlabs.com/docs/v23.2/node-shutdown?filters=decommission#decommission-the-node

> - Fixed a bug introduced in v22.2 where adding multiple columns with unique constraints in a single statement could result in the error `secondary index for backfill contains physical column not present in source primary index`. [#118137][#118137]

Reviewer: 'unique constraints' could link to https://www.cockroachlabs.com/docs/v23.2/unique

- Fixed a bug where an `unable to get CPU capacity` error could be logged every 10 seconds when running outside of a CPU `cgroup`. [#118670][#118670]

> - Fixed a bug where an `AUTO CREATE STATS` job could cause an internal system table to grow, and this could cause slow queries related to the job system. [#118944][#118944]

Reviewer: 'AUTO CREATE STATS job' could link to https://www.cockroachlabs.com/docs/v23.2/show-jobs#show-automatic-jobs

Reviewer: PS my recent links should have been to v23.1 docs, sorry. Our search results default to v23.2, oops

> - Fixed a bug that caused a confusing error when a sequence name allocated by `SERIAL` conflicted with an existing type name. [#118948][#118948]

Reviewer: SERIAL could link to https://www.cockroachlabs.com/docs/v23.1/serial

> - Fixed a bug where an `ALTER PRIMARY KEY` query could fail with the error `non-nullable column {x} with no value! Index scanned ..` when validating a recreated secondary index. [#118972][#118972]

Reviewer: ALTER PRIMARY KEY could link to https://www.cockroachlabs.com/docs/v23.1/alter-table#alter-primary-key

## v23.1.17

Release Date: March 19, 2024

{% include releases/release-downloads-docker-image.md release=include.release %}

<h3 id="v23-1-17-security-updates">Security updates</h3>

> - A user with the `VIEWACTIVITY` privilege can now request a statement bundle from the [**Statements** page](https://www.cockroachlabs.com/docs/v23.1/ui-statements-page) in the DB Console or with the internal builtin function `crdb_internal.requets_statement_bundle`. [#119638][#119638]

Author: ```suggestion
- A user with the `VIEWACTIVITY` privilege can now request a statement bundle from the [**Statements** page](https://www.cockroachlabs.com/docs/v23.1/ui-statements-page) in the DB Console or with the internal builtin function `crdb_internal.requests_statement_bundle`. [#119638][#119638]
```

> - Clusters using [Cluster Single Sign-on (SSO) using JSON web tokens (JWTs)](https://www.cockroachlabs.com/docs/v23.1/sso-sql) can now optionally fetch signing keys from configured issuers instead of configuring static signing keys for each issuer. When new cluster setting `server.jwt_authentication.jwks_auto_fetch.enabled` is set to `true`, signing keys are automatically fetched from the issuer using metadata published in its OpenID configuration. In this case, static signing keys in `server.jwt_authentication.jwks` are ignored. When automatch fetching is enabled, there may be a slight increase network latency in each JWT authentication request, proportional to the latency between the cluster and the issuer's endpoint. [#120063][#120063]

Author: ```suggestion
- Clusters using [Cluster Single Sign-on (SSO) with JSON web tokens (JWTs)](https://www.cockroachlabs.com/docs/v23.1/sso-sql) can now optionally fetch signing keys from configured issuers instead of configuring static signing keys for each issuer. When the new cluster setting `server.jwt_authentication.jwks_auto_fetch.enabled` is set to `true`, signing keys are automatically fetched from the issuer using metadata published in its OpenID configuration. In this case, static signing keys in `server.jwt_authentication.jwks` are ignored. When automatic fetching is enabled, there may be a slight increase in network latency for each JWT authentication request, proportional to the latency between the cluster and the issuer's endpoint. [#120063][#120063]
```

- Added support for [index hinting](https://www.cockroachlabs.com/docs/v23.1/table-expressions#force-index-selection) with `INSERT` and `UPSERT` statements. This allows `INSERT ... ON CONFLICT` and `UPSERT` queries to use index hints in the same way as `UPDATE` and `DELETE` statements. [#119600][#119600]

<h3 id="v23-1-17-operational-changes">Operational changes</h3>

- You can now pass the `--include-range-info` flag when [generating a diagnostic bundle](https://www.cockroachlabs.com/docs/v23.1/cockroach-debug-zip) to include problem ranges. [#119233][#119233]
- In unredacted diagnostic bundles, two columns have been added to the `crdb_internal.transaction_contention_events` table:
    - `waiting_stmt_query`: the query of the statement that is waiting to run.
    - `blocking_txn_queries_unordered`: an unordered list of the blocking transaction's queries. [#119422]

>   [#119422]

Author: ```suggestion
    [#119422]
```

<h3 id="v23-1-17-db-console-changes">DB Console changes</h3>

- The [**Statements** page](https://www.cockroachlabs.com/docs/v23.1/ui-statements-page) now always shows the entire selected period, instead of only the period that contains data. [#118811][#118811]
- The [**Overload Dashboard**](https://www.cockroachlabs.com/docs/v23.1/ui-overload-dashboard) page now includes two additional graphs:
    - **Elastic CPU Utilization**: Shows the actual CPU used for elastic work compared with the configured limit.
    - **Elastic CPU Exhausted Duration Per Second**: Shows how much time, in milliseconds, that elastic work has been subject to CPU exhaustion. [#118908][#118908]

<h3 id="v23-1-17-bug-fixes">Bug fixes</h3>

- Fixed a bug where creating a [changefeed](https://www.cockroachlabs.com/docs/v23.1/change-data-capture-overview) that targeted tables with a zero-scale `DECIMAL(n)` column, `format='avro'`, and `diff` would cause a panic. [#118848][#118848]

> - - Fixed a bug where a [changefeed](https://www.cockroachlabs.com/docs/v23.1/change-data-capture-overview.html) could omit events in rare cases, logging the error `cdc ux violation: detected timestamp ... that is less or equal to the local frontier`. This could happen in the following scenario:

Author: ```suggestion
- Fixed a bug where a [changefeed](https://www.cockroachlabs.com/docs/v23.1/change-data-capture-overview.html) could omit events in rare cases, logging the error `cdc ux violation: detected timestamp ... that is less or equal to the local frontier`. This could happen in the following scenario:
```

- Fixed a bug that caused a confusing error when a sequence name allocated by `SERIAL` conflicted with an existing type name. [#118948][#118948]
- Fixed a bug where an `ALTER PRIMARY KEY` query could fail with the error `non-nullable column {x} with no value! Index scanned ..` when validating a recreated secondary index. [#118972][#118972]
- Fixed a bug introduced in v23.1 that could cause the internal error `attempting to append refresh spans after the tracked timestamp has moved forward` to be logged in some cases when using virtual tables such as`crdb_internal.system_jobs`. [#119186][#119186]
- Fixed a bug in the `crdb_internal.leases` table where a deadlock in the leasing system could cause a node to become unavailable. [#119374][#119374]
- Fixed a bug introduced in v23.1.15 where a rangefeed-resolved timestamp could get stuck after a range merge if the range cache thinks the barrier spans multiple ranges. The log message `pushing old intents failed: range barrier failed, range split` is constantly emitted. [#119543][#119543]

> - Fixed a rare panic that could happen during a `pg_dump` import that contains a subquery in one of its argume nts, such as a function like `SELECT addgeometrycolumn(...)`. Now, attempting to import a `pg_dump` with a function that has a subquery in one of its arguments results in an error rather than a panic. [#118611][#118611]

Author: ```suggestion
- Fixed a rare panic that could happen during a `pg_dump` import that contains a subquery in one of its arguments, such as a function like `SELECT addgeometrycolumn(...)`. Now, attempting to import a `pg_dump` with a function that has a subquery in one of its arguments results in an error rather than a panic. [#118611][#118611]
```

[#118992]: https://github.com/cockroachdb/cockroach/pull/118992
[#119186]: https://github.com/cockroachdb/cockroach/pull/119186
[#119233]: https://github.com/cockroachdb/cockroach/pull/119233
[#119262]: https://github.com/cockroachdb/cockroach/pull/119262
[#119338]: https://github.com/cockroachdb/cockroach/pull/119338
[#119374]: https://github.com/cockroachdb/cockroach/pull/119374
[#119400]: https://github.com/cockroachdb/cockroach/pull/119400
[#119422]: https://github.com/cockroachdb/cockroach/pull/119422
[#119466]: https://github.com/cockroachdb/cockroach/pull/119466
[#119495]: https://github.com/cockroachdb/cockroach/pull/119495
[#119543]: https://github.com/cockroachdb/cockroach/pull/119543
[#119600]: https://github.com/cockroachdb/cockroach/pull/119600
[#119634]: https://github.com/cockroachdb/cockroach/pull/119634
[#119638]: https://github.com/cockroachdb/cockroach/pull/119638
[#119695]: https://github.com/cockroachdb/cockroach/pull/119695
[#119712]: https://github.com/cockroachdb/cockroach/pull/119712
[#120063]: https://github.com/cockroachdb/cockroach/pull/120063
[#120244]: https://github.com/cockroachdb/cockroach/pull/120244

> [38993cb9c]: https://github.com/cockroachdb/cockroach/commit/38993cb9c
> [566a30300]: https://github.com/cockroachdb/cockroach/commit/566a30300
> [61776d5f5]: https://github.com/cockroachdb/cockroach/commit/61776d5f5

Author: ```suggestion
```
